bentoml/_internal/service/inference_api.py

```diff
@@ -39,7 +45,77 @@ def __init__(
+            for key in input_type:
+                if key not in sig.parameters:
+                    if (
+                        key in ["context", "ctx"]
+                        or sig.parameters[key].annotation == bentoml.Context
+                    ):
+                        if self.needs_ctx:
+                            raise ValueError(
+                                f"API function has two context parameters: '{self.ctx_param}' and '{key}'; it should only have one."
+                            )
+
+                        self.needs_ctx = True
+                        self.ctx_param = key
+                        continue
+
+                    raise ValueError(
+                        f"API function has extra parameter with name '{key}'."
+                    )
+
+                annotation: t.Type[t.Any] = sig.parameters[key].annotation
+                if (
+                    isinstance(annotation, t.Type)
+                    and annotation != inspect.Signature.empty
+                ):
+                    # if type annotations have been successfully resolved
+                    if not is_compatible_type(input_type[key], annotation):
+                        raise TypeError(
+                            f"Expected type of argument '{key}' to be '{input_type[key]}', got '{sig.parameters[key].annotation}'"
+                        )
+
+            expected_args = len(input_type) + (1 if self.needs_ctx else 0)
+            if len(sig.parameters) != expected_args:
+                raise ValueError(
+                    f"expected API function to have arguments ({', '.join(input_type.keys())}, [context]), got ({', '.join(sig.parameters.keys())})"
+                )
+
+        else:
+            param_iter = iter(sig.parameters)
+            first_arg = next(param_iter)
+            annotation = sig.parameters[first_arg].annotation
+            if isinstance(annotation, t.Type) and annotation != inspect.Signature.empty:
+                if not is_compatible_type(input_type, annotation):
+                    raise TypeError(
+                        f"Expected type of argument '{first_arg}' to be '{input_type}', got '{sig.parameters[first_arg].annotation}'"
+                    )
+
+            if len(sig.parameters) > 2:
+                raise ValueError("API function should only take one or two arguments")
+            elif len(sig.parameters) == 2:
+                self.needs_ctx = True
+
+                second_arg = next(param_iter)
+                annotation = sig.parameters[second_arg].annotation
+                if (
+                    isinstance(annotation, t.Type)
+                    and annotation != inspect.Signature.empty
+                ):
+                    if not annotation == bentoml.Context:
+                        raise TypeError(
+                            f"Expected type of argument '{second_arg}' to be '{input_type}', got '{sig.parameters[second_arg].annotation}'"
+                        )
```

Comment:
Seems like we are enforcing the `context` argument to be the last argument is in the singular input case but not the multi-input case. Is my understanding correct? If yes, what is the reason behind?

Reply:
I think this was purely because I didn't want to introduce another function call as overhead by adding a `partial` call here since this is technically on the critical serving path.

At some point I think we could factor the if statements out of the serving loop too, but it's probably worth getting benchmarks actually setup first.

Reply:
The code here is not, I was talking about the nested ifs in the server code.